src/lib.rs

```diff
@@ -140,19 +174,41 @@ pub fn run(args: Args) {
             );
 
             printer.get_mut().clear();
+            let mut sink = printer.sink_with_path(path);
             get_searcher(&args)
                 .borrow_mut()
-                .search_path(query_context, path, printer.sink_with_path(path))
+                .search_path(query_context, path, &mut sink)
                 .unwrap();
+            if sink.has_match() {
+                matched.store(true, Ordering::SeqCst);
+            }
             buffer_writer.print(printer.get_mut()).unwrap();
         });
 
-    cached_queries.error_if_no_successful_query_parsing();
+    if !messages::errored() {
+        if !searched.load(Ordering::SeqCst) {
+            eprint_nothing_searched();
+        } else {
+            cached_queries.error_if_no_successful_query_parsing();
+        }
+    }
+
+    if messages::errored() {
+        process::exit(2);
```

Comment:
This mimics `ripgrep`, I guess it uses exit status 2 to indicate "actual failure" and 1 to indicate "no matches found"

Reply:
Maybe worth making an `ExitCode` enum (and then I guess your own `exit` function) to make that clear?